src/test/java/unit/CloneWorkspace.java

```diff
@@ -0,0 +1,200 @@
+      ExternalBQDatasets.deleteDataset(externalDataset);
+      externalDataset = null;
+    }
+  }
+
+  @Test
+  public void cloneWorkspace(TestInfo testInfo) throws Exception {
+    workspaceCreator.login();
+
+    // create a workspace
+    // `terra workspace create --format=json`
+    sourceWorkspace =
+        TestCommand.runAndParseCommandExpectSuccess(UFWorkspace.class, "workspace", "create");
+
+    // Switch to the new workspace
+    // `terra workspace set --id=$id`
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + sourceWorkspace.id);
+
+    // Add a bucket resource
+    UFGcsBucket sourceBucket =
+        TestCommand.runAndParseCommandExpectSuccess(
+            UFGcsBucket.class,
+            "resource",
+            "create",
+            "gcs-bucket",
+            "--name=" + "bucket_1",
+            "--bucket-name=" + UUID.randomUUID(),
+            "--cloning=COPY_RESOURCE");
+
+    // Add another bucket resource with COPY_NOTHING
+    UFGcsBucket copyNothingBucket =
+        TestCommand.runAndParseCommandExpectSuccess(
+            UFGcsBucket.class,
+            "resource",
+            "create",
+            "gcs-bucket",
+            "--name=" + "bucket_2",
+            "--bucket-name=" + UUID.randomUUID(),
+            "--cloning=COPY_NOTHING");
+
+    // Add a dataset resource
+    UFBqDataset sourceDataset =
+        TestCommand.runAndParseCommandExpectSuccess(
+            UFBqDataset.class,
+            "resource",
+            "create",
+            "bq-dataset",
+            "--name=dataset_1",
+            "--dataset-id=dataset_1",
+            "--description=The first dataset.",
+            "--cloning=COPY_RESOURCE");
+
+    // Add a referenced resource
+    externalDataset = ExternalBQDatasets.createDataset();
+    ExternalBQDatasets.grantReadAccess(
+        externalDataset, workspaceCreator.email, ExternalBQDatasets.IamMemberType.USER);
+
+    // grant the user's proxy group access to the dataset so that it will pass WSM's access check
+    // when adding it as a referenced resource
+    ExternalBQDatasets.grantReadAccess(
+        externalDataset, Auth.getProxyGroupEmail(), ExternalBQDatasets.IamMemberType.GROUP);
+
+    UFBqDataset datasetReference =
+        TestCommand.runAndParseCommandExpectSuccess(
+            UFBqDataset.class,
+            "resource",
+            "add-ref",
+            "bq-dataset",
+            "--name=" + "dataset_ref",
+            "--project-id=" + externalDataset.getProjectId(),
+            "--dataset-id=" + externalDataset.getDatasetId(),
+            "--cloning=COPY_REFERENCE");
+
+    // Clone the workspace
+    UFClonedWorkspace clonedWorkspace =
+        TestCommand.runAndParseCommandExpectSuccess(
+            UFClonedWorkspace.class,
+            "workspace",
+            "clone",
+            "--name=" + "cloned_workspace",
+            "--description=A clone.");
+
+    assertEquals(sourceWorkspace.id, clonedWorkspace.sourceWorkspace.id);
```

Comment:
please include a message for each assertion. I find it makes it much easier to track down a test failure, since the line numbers don't always match up.

Reply:
👍🏻